saleor/order/utils.py

```diff
@@ -342,10 +342,41 @@ def match_orders_with_new_user(user: User) -> None:
     Order.objects.confirmed().filter(user_email=user.email, user=None).update(user=user)
 
 
+def _chunk_products(products, product_limit):
+    chunks = []
+    for i in range(0, len(products), product_limit):
+        limit = i + product_limit
+        chunks.append(products[i:limit])
+    return chunks
+
+
+def _get_page_count(products_first_page, rest_of_products):
+    if not rest_of_products and len(products_first_page) == 4:
+        return 2
+    try:
+        append_last_page = 1 if len(rest_of_products[-1]) > 11 else 0
+        return 1 + len(rest_of_products) + append_last_page
+    except IndexError:
+        return 1
+
+
 def generate_invoice_pdf_for_order(invoice):
     logo_path = static_finders.find("images/logo-light.svg")
+
+    all_products = invoice.order.lines.all()
+    _product_limit = 3 if len(all_products) < 4 else 4
+    products_first_page = all_products[:_product_limit]
+    rest_of_products = _chunk_products(all_products[_product_limit:], 13)
```

Comment:
There is a lot of magic numbers here; I'm wondering if we can come up with some meaningful names for them (at least for some of them) and make them constants. E.g. this `13` could become something as `MAX_PRODUCTS_PER_PAGE` and the number `4` above `MAX_PRODUCTS_FIRST_PAGE`, what do you think?

Reply:
Good idea, I know this logic is quite enigmatic, will try to make it more explicit.